editoast/src/documents.rs

```diff
@@ -0,0 +1,267 @@
+//! This module manage documents in the database.
+//! You can add, retrieve and delete documents.
+//!
+//! Each document has a unique key that you have to provide or build.
+//!
+//! ## Example
+//!
+//! ```rust
+//! let study_pk = 42;
+//! let key = DocumentKey::new_hash(&format!("study_image_{study_pk}"));
+//! let image: Vec<u8> = ...;
+//! let doc = Document::insert(db_pool, &key, ContentType::png(), image).await?;
+//! let document_pk = doc.get_pk();
+//! let image_url = ket.get_url("http://localhost:8090");
+//! Document::delete(db_pool, &key).await?;
+//! ```
+
+use crate::error::{EditoastError, Result};
+use crate::tables::osrd_infra_document;
+use crate::DbPool;
+use actix_web::http::header::ContentType;
+use actix_web::web::{block, Data};
+use diesel::expression_methods::ExpressionMethods;
+use diesel::result::{DatabaseErrorKind, Error as DieselError};
+use diesel::{insert_into, QueryDsl, RunQueryDsl};
+use editoast_derive::EditoastError;
+use sha1::{Digest, Sha1};
+use thiserror::Error;
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "documents")]
+enum DocumentErrors {
+    #[error("Document not found")]
+    #[editoast_error(status = 404)]
+    NotFound,
+    #[error("A document with the same key already exists")]
+    #[editoast_error(status = 500)]
+    AlreadyExists,
+}
+
+#[derive(Debug, Insertable)]
+#[diesel(table_name = osrd_infra_document)]
+struct DocumentForm {
+    key: String,
+    content_type: String,
+    data: Vec<u8>,
+}
+
+/// Represent a document key.
+/// Can be created manually or build from hashable objects.
+///
+/// **WARNING**: The key of a document must be unique.
+///
+/// ## Example
+///
+/// Let's say we have an image of a study to save in the database.
+/// A way to construct your unique key can be to use a tuple of the study id and a static `str`.
+///
+/// ```
+/// let study_pk = 42;
+/// let key = DocumentKey::new_hash(&format("study_image_{study_pk}"));
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DocumentKey(String);
+
+impl DocumentKey {
+    /// Create a document key from a string
+    pub fn new<T: AsRef<str>>(key: T) -> Self {
+        Self(key.as_ref().into())
+    }
+
+    /// Create a document key from any data hashable object
+    #[allow(dead_code)] // TODO: Remove once it's used by studies or rolling stocks
+    pub fn new_hash<T: AsRef<[u8]>>(hashable_objects: &T) -> Self {
+        let mut hasher = Sha1::new();
```

Comment:
is it intended to be a cryptographic hash?

Reply:
I don't especially need a cryptographic hash. I just need a hashing algorithm that has strong collision resistance. Sha1 has this property and is fast to compute. Do you have another algorithm in mind?